crates/accelerate/src/inverse_cancellation.rs

```diff
@@ -0,0 +1,193 @@
+use indexmap::IndexMap;
+use pyo3::prelude::*;
+use rustworkx_core::petgraph::stable_graph::NodeIndex;
+
+use qiskit_circuit::circuit_instruction::OperationFromPython;
+use qiskit_circuit::dag_circuit::{DAGCircuit, NodeType};
+use qiskit_circuit::operations::Operation;
+use qiskit_circuit::packed_instruction::PackedInstruction;
+
+fn gate_eq(py: Python, gate_a: &PackedInstruction, gate_b: &OperationFromPython) -> PyResult<bool> {
+    if gate_a.op.name() != gate_b.operation.name() {
+        return Ok(false);
+    }
+    let a_params = gate_a.params_view();
+    if a_params.len() != gate_b.params.len() {
+        return Ok(false);
+    }
+    let mut param_eq = true;
+    for (a, b) in a_params.iter().zip(&gate_b.params) {
+        if !a.is_close(py, b, 1e-10)? {
+            param_eq = false;
+            break;
+        }
+    }
+    Ok(param_eq)
+}
+
+fn run_on_self_inverse(
+    py: Python,
+    dag: &mut DAGCircuit,
+    op_counts: &IndexMap<String, usize, RandomState>,
+    self_inverse_gate_names: HashSet<String>,
+    self_inverse_gates: Vec<OperationFromPython>,
+) -> PyResult<()> {
+    if !self_inverse_gate_names
+        .iter()
+        .any(|name| op_counts.contains_key(name))
+    {
+        return Ok(());
+    }
+    for gate in self_inverse_gates {
+        let gate_count = op_counts.get(gate.operation.name()).unwrap_or(&0);
+        if *gate_count <= 1 {
+            continue;
+        }
+        let mut collect_set: HashSet<String> = HashSet::with_capacity(1);
+        collect_set.insert(gate.operation.name().to_string());
+        let gate_runs: Vec<Vec<NodeIndex>> = dag.collect_runs(collect_set).unwrap().collect();
+        for gate_cancel_run in gate_runs {
+            let mut partitions: Vec<Vec<NodeIndex>> = Vec::new();
+            let mut chunk: Vec<NodeIndex> = Vec::new();
+            let max_index = gate_cancel_run.len() - 1;
+            for (i, cancel_gate) in gate_cancel_run.iter().enumerate() {
+                let node = &dag.dag[*cancel_gate];
+                if let NodeType::Operation(inst) = node {
+                    if gate_eq(py, inst, &gate)? {
+                        chunk.push(*cancel_gate);
+                    } else {
+                        let is_empty: bool = chunk.is_empty();
+                        if !is_empty {
+                            partitions.push(std::mem::take(&mut chunk));
+                        }
+                        continue;
+                    }
+                    if i == max_index {
+                        partitions.push(std::mem::take(&mut chunk));
+                    } else {
+                        let next_qargs = if let NodeType::Operation(next_inst) =
+                            &dag.dag[gate_cancel_run[i + 1]]
+                        {
+                            next_inst.qubits
+                        } else {
+                            panic!("Not an op node")
+                        };
+                        if inst.qubits != next_qargs {
```

Comment:
Cool! I think this is the first example I've seen of the intern IDs saving us some time :)